File: backend/models/historyModel.js
const {DataTypes} = require("sequelize");
const sequelize = require("../config/database");
const LeadStatus = require("../models/leadStatusModel");
const Employee = require("../models/employeesModel");

const History = sequelize.define(
    "History",
    {
        history_id : {
            type : DataTypes.INTEGER,
            primaryKey : true,
            autoIncrement : true
        },
        lead_id :{
            type : DataTypes.INTEGER,
            allowNull : false
        },
        next_meeting:{
            type : DataTypes.DATE
        },
        status_id:{
            type: DataTypes.INTEGER,
            allowNull: false
        },
        previous_status_id: {
            type: DataTypes.INTEGER
        },
        changed_by_emp_id: {
            type: DataTypes.STRING
        },
        loanType: {
            type: DataTypes.STRING
        },
        remark: {
            type: DataTypes.STRING
        }
    },
    {
    tableName: "histories",
    timestamps: true,
  }
);

History.belongsTo(LeadStatus, {
    foreignKey: "status_id",
    as: "statusDetails",
});

History.belongsTo(LeadStatus, {
    foreignKey: "previous_status_id",
    as: "previousStatusDetails",
});

History.belongsTo(Employee, {
    foreignKey: "changed_by_emp_id",
    as: "changedBy",
});

    History.sync({alter : false}).then(()=> {
        console.log("History table created")
    }).catch((error) => {
        console.error("Error creating history table", error);
    });

module.exports = History;

File: frontend/lib/models/history_model.dart
class History {
  final history_id;
  final lead_id;
  final changedBy;
  final changedByEmpId;
  final previousStatus;
  final status;
  final next_meeting;
  final loanType;
  final remark;
  final createdAt;

  History({
    this.history_id,
    this.lead_id,
    this.changedBy,
    this.changedByEmpId,
    this.previousStatus,
    this.status,
    this.next_meeting,
    this.loanType,
    this.remark,
    this.createdAt,
  });

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> json = {};
    if (lead_id != null) json["lead_id"] = lead_id;
    if (changedByEmpId != null) json["changed_by_emp_id"] = changedByEmpId;
    if (next_meeting != null) json["next_meeting"] = next_meeting;
    if (status != null) json["status"] = status;
    if (loanType != null) json["loanType"] = loanType;
    if (remark != null) json["remark"] = remark;
    return json;
  }

  factory History.fromJson(Map<String, dynamic> json) {
    return History(
      history_id: json['history_id'],
      lead_id: json['lead_id'],
      changedBy: json['changedBy'] != null ? json['changedBy']['ename'] : json['changed_by_emp_id'] ?? "",
      changedByEmpId: json['changedBy'] != null ? json['changedBy']['emp_id'] : json['changed_by_emp_id'] ?? "",
      previousStatus: json['previousStatusDetails'] != null ? json['previousStatusDetails']['name'] : json['previous_status_id']?.toString() ?? "",
      status: json['statusDetails'] != null ? json['statusDetails']['name'] : json['status'] ?? "",
      next_meeting: json['next_meeting'] ?? "",
      loanType: json['loanType'] ?? "",
      remark: json['remark'] ?? "",
      createdAt: json['createdAt'] ?? "",
    );
  }
}
